### PR TITLE
8277042: add test for 8276036 to compiler/codecache

### PR DESCRIPTION
**Pull request by tkiriyama**
Could you please review the 8277042 code?  
This is the enhancement for 8276036.  
I add a new test to verify the value of full_count in the message of insufficient codecache.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8277042](https://bugs.openjdk.java.net/browse/JDK-8277042): add test for 8276036 to compiler/codecache

### Reviewers
 * [Christian Hagedorn](https://openjdk.java.net/census#chagedorn) (@chhagedorn - **Reviewer**) ⚠️ Review applies to 4d29d81a11d1b70ba546f13f149a81eb907a39f1
 * [Tobias Hartmann](https://openjdk.java.net/census#thartmann) (@TobiHartmann - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/6364/head:pull/6364` \
`$ git checkout pull/6364`

Update a local copy of the PR: \
`$ git checkout pull/6364` \
`$ git pull https://git.openjdk.java.net/jdk pull/6364/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 6364`

View PR using the GUI difftool: \
`$ git pr show -t 6364`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/6364.diff">https://git.openjdk.java.net/jdk/pull/6364.diff</a>

</details>

### CODE REVIEW

File: test/hotspot/jtreg/compiler/codecache/CodeCacheFullCountTest.java [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+            } catch (ClassNotFoundException e) {
+                return Class.forName(name, resolve, CodeCacheFullCountTest.class.getClassLoader());
+            }
+        }
+    }
+}
+
+abstract class Foo {
+    public abstract int foo();
+}
```
Comment (chhagedorn) [inline test/hotspot/jtreg/compiler/codecache/CodeCacheFullCountTest.java:97]: Add some spaces between the classes to improve the readability.

File: test/hotspot/jtreg/compiler/codecache/CodeCacheFullCountTest.java [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+    public static void wasteCodeCache()  throws Exception {
+        URL url = CodeCacheFullCountTest.class.getProtectionDomain().getCodeSource().getLocation();
+
+        for (int i = 0; i < 500; i++) {
+            ClassLoader cl = new MyClassLoader(url);
+            refClass(cl.loadClass("SomeClass"));
+        }
+    }
+
+    public static void runtest() throws Throwable {
```
Comment (chhagedorn) [inline test/hotspot/jtreg/compiler/codecache/CodeCacheFullCountTest.java:56]: Should be CamelCase: `runTest()`.

File: test/hotspot/jtreg/compiler/codecache/CodeCacheFullCountTest.java [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+
+        Pattern pattern = Pattern.compile("full_count=(\\d)");
+        Matcher stdoutMatcher = pattern.matcher(stdout);
+        if (stdoutMatcher.find()) {
+            int fullCount = Integer.parseInt(stdoutMatcher.group(1));
+            if (fullCount != 1) {
+                throw new RuntimeException("the value of full_count is wrong.");
+            }
+        } else {
+            throw new RuntimeException("codecache shortage was not occured.");
```
Comment (chhagedorn) [inline test/hotspot/jtreg/compiler/codecache/CodeCacheFullCountTest.java:71]: `was not occured` -> `did not occur`

**Review by chhagedorn — APPROVED** [review]
Otherwise, looks good! Thanks for adding this test.

**Comment by tkiriyama** [comment]
Hi, @chhagedorn . Thank you for your comment.
I fixed this test as per your instructions.

**Review by TobiHartmann — APPROVED** [review]
Looks good to me.

**Comment by chhagedorn** [comment]
I ran your test with latest JDK and found the following crash:
```
#
# A fatal error has been detected by the Java Runtime Environment:
#
#  Internal Error (/opt/mach5/mesos/work_dir/slaves/ff806ead-2cac-495d-9cbc-62116f99bf14-S14125/frameworks/1735e8a2-a1db-478c-8104-60c8b0af87dd-0196/executors/928f8391-f7c3-41f7-944d-47d10f56f83d/runs/8280836e-2a4d-4ec6-8bf2-5d87f2a87640/workspace/open/src/hotspot/share/runtime/mutex.cpp:444), pid=16467, tid=16482
#  assert(false) failed: Attempting to acquire lock CompileTask_lock/safepoint out of order with lock MethodCompileQueue_lock/safepoint -- possible deadlock
#
# JRE version: Java(TM) SE Runtime Environment (18.0) (fastdebug build 18-internal+0-2021-11-16-1108284.christian.hagedorn.jdk)
# Java VM: Java HotSpot(TM) 64-Bit Server VM (fastdebug 18-internal+0-2021-11-16-1108284.christian.hagedorn.jdk, mixed mode, sharing, tiered, compressed oops, compressed class ptrs, g1 gc, linux-amd64)
# Problematic frame:
# V  [libjvm.so+0x1563e4d]  Mutex::check_rank(Thread*)+0x29d
```

I will have a closer look and get back to you again. It seems that your test revealed an existing bug. 

Please do not sponsor this PR, yet!

**Comment by chhagedorn** [comment]
I could also trigger this locally. I filed https://bugs.openjdk.java.net/browse/JDK-8277213. This bug should be fixed before this test can be integrated to avoid noise in our testing.

File: test/hotspot/jtreg/compiler/codecache/CodeCacheFullCountTest.java [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+import java.net.URLClassLoader;
+import java.lang.reflect.Field;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+import jdk.test.lib.process.OutputAnalyzer;
+import jdk.test.lib.process.ProcessTools;
+
+/*
+ * @test
+ * @bug 8276036
```
Comment (TobiHartmann) [inline test/hotspot/jtreg/compiler/codecache/CodeCacheFullCountTest.java:34]: I have a fix ready for JDK-8277213 (see https://github.com/openjdk/jdk/pull/6449). Could you please add the bug ID to this test?

**Comment by TobiHartmann** [comment]
Now even with the fix for https://github.com/openjdk/jdk/pull/6449, I hit an assert with the new test once that I'm unfortunately not able to reproduce even after hundreds of runs:

```
# A fatal error has been detected by the Java Runtime Environment:
#
#  Internal Error (/workspace/open/src/hotspot/share/compiler/compileBroker.cpp:369), pid=25796, tid=6403
#  assert(_last->next() == __null) failed: not last

Stack: [0x000070000838a000,0x000070000848a000],  sp=0x00007000084862f0,  free space=1008k
Native frames: (J=compiled Java code, j=interpreted, Vv=VM code, C=native code)
V  [libjvm.dylib+0x1159a39]  VMError::report_and_die(int, char const*, char const*, __va_list_tag*, Thread*, unsigned char*, void*, void*, char const*, int, unsigned long)+0x6e9
V  [libjvm.dylib+0x115a0bb]  VMError::report_and_die(Thread*, void*, char const*, int, char const*, char const*, __va_list_tag*)+0x3b
V  [libjvm.dylib+0x628fbd]  report_vm_error(char const*, int, char const*, char const*, ...)+0xdd
V  [libjvm.dylib+0x5e394a]  CompileQueue::add(CompileTask*)+0x8a
V  [libjvm.dylib+0x5e718a]  CompileBroker::compile_method_base(methodHandle const&, int, int, methodHandle const&, int, CompileTask::CompileReason, bool, Thread*)+0xa3a
V  [libjvm.dylib+0x5e7f5d]  CompileBroker::compile_method(methodHandle const&, int, int, methodHandle const&, int, CompileTask::CompileReason, DirectiveSet*, JavaThread*)+0x6ad
V  [libjvm.dylib+0x5e788e]  CompileBroker::compile_method(methodHandle const&, int, int, methodHandle const&, int, CompileTask::CompileReason, JavaThread*)+0xbe
V  [libjvm.dylib+0x5c744e]  CompilationPolicy::compile(methodHandle const&, int, CompLevel, JavaThread*)+0x4ee
V  [libjvm.dylib+0x5c69dc]  CompilationPolicy::event(methodHandle const&, methodHandle const&, int, int, CompLevel, CompiledMethod*, JavaThread*)+0x20c
V  [libjvm.dylib+0x8eea04]  InterpreterRuntime::frequency_counter_overflow_inner(JavaThread*, unsigned char*)+0x304
V  [libjvm.dylib+0x8ee4ca]  InterpreterRuntime::frequency_counter_overflow(JavaThread*, unsigned char*)+0x1a
j  Foo3.foo()I+0
j  SomeClass.<clinit>()V+61
v  ~StubRoutines::call_stub
``` 

It seems to be a concurrency issue. I filed [JDK-8277441](https://bugs.openjdk.java.net/browse/JDK-8277441) to track this and will set up a server machine to run it for several days. I'll report back but until then, we should still hold of with integrating this or integrate and problem list right away.

@tkiriyama Good work with the test which already caught two bugs!

File: test/hotspot/jtreg/compiler/codecache/CodeCacheFullCountTest.java (hunk 4 above)
Comment (tkiriyama) [reply test/hotspot/jtreg/compiler/codecache/CodeCacheFullCountTest.java]: I see. I added the bug ID to this test for JDK-8277213.
I will probably add JDK-8277441 currently being discussed.

Comment (TobiHartmann) [reply test/hotspot/jtreg/compiler/codecache/CodeCacheFullCountTest.java]: Thanks, yes, please add 8277441 as well.

Comment (TobiHartmann) [reply test/hotspot/jtreg/compiler/codecache/CodeCacheFullCountTest.java]: I just pushed JDK-8277441 and verified that the test now always passed.

Comment (tkiriyama) [reply test/hotspot/jtreg/compiler/codecache/CodeCacheFullCountTest.java]: Thank you. I'm glad the bug was fixed.
I added 8277441 to this test.

**Review by TobiHartmann — APPROVED** [review]
Thanks, looks good.

**Pull request closed without merging**